src/Dialect/ONNX/ONNXOps/Math/Clip.cpp

```diff
@@ -41,46 +59,80 @@ LogicalResult ONNXClipOpShapeHelper::computeShape() {
+    Value max, Operation *op) {
   // Look at input.
-  if (!hasShapeAndRank(getInput()))
+  if (!hasShapeAndRank(input))
     return success();
-  RankedTensorType inputTy = getInput().getType().cast<RankedTensorType>();
+  RankedTensorType inputTy = input.getType().cast<RankedTensorType>();
   Type elementType = inputTy.getElementType();
   // Look at optional min.
-  if (!getMin().getType().isa<NoneType>()) {
+  if (!min.getType().isa<NoneType>()) {
     // Has a min, make sure its of the right type.
-    if (!hasShapeAndRank(getMin()))
+    if (!hasShapeAndRank(min))
       return success();
     // And size.
-    RankedTensorType minTy = getMin().getType().cast<RankedTensorType>();
+    RankedTensorType minTy = min.getType().cast<RankedTensorType>();
     if (minTy.getElementType() != elementType)
-      return emitError("Element type mismatch between input and min tensors");
+      return op->emitError(
+          "Element type mismatch between input and min tensors");
     if (minTy.getShape().size() != 0)
-      return emitError("Min tensor ranked with nonzero size");
+      return op->emitError("Min tensor ranked with nonzero size");
   }
   // Look at optional max
-  if (!getMax().getType().isa<NoneType>()) {
+  if (!max.getType().isa<NoneType>()) {
     // Has a max, make sure its of the right type.
-    if (!hasShapeAndRank(getMax()))
+    if (!hasShapeAndRank(max))
       return success();
     // And size.
-    RankedTensorType maxTy = getMax().getType().cast<RankedTensorType>();
+    RankedTensorType maxTy = max.getType().cast<RankedTensorType>();
     if (maxTy.getElementType() != elementType)
-      return emitError("Element type mismatch between input and max tensors");
+      return op->emitError(
+          "Element type mismatch between input and max tensors");
     if (maxTy.getShape().size() != 0)
-      return emitError("Min tensor ranked with nonzero size");
+      return op->emitError("Min tensor ranked with nonzero size");
   }
+  ONNXNonSpecificOpShapeHelper<T> shapeHelper(op, {});
+  return shapeHelper.computeShapeAndUpdateType(elementType);
+}
 
-  ONNXClipOpShapeHelper shapeHelper(getOperation(), {});
+LogicalResult ONNXClipOp::inferShapes(
+    std::function<void(Region &)> doShapeInference) {
+  return ONNXClipOpGeneralInferShapes<ONNXClipOp>(
+      doShapeInference, getInput(), getMin(), getMax(), getOperation());
+}
+
+LogicalResult ONNXClipV6Op::inferShapes(
+    std::function<void(Region &)> doShapeInference) {
+  if (!hasShapeAndRank(getInput()))
+    return success();
+  RankedTensorType inputTy = getInput().getType().cast<RankedTensorType>();
+  Type elementType = inputTy.getElementType();
+  ONNXClipV6OpShapeHelper shapeHelper(getOperation(), {});
   return shapeHelper.computeShapeAndUpdateType(elementType);
 }
 
+LogicalResult ONNXClipV11Op::inferShapes(
+    std::function<void(Region &)> doShapeInference) {
+  return ONNXClipOpGeneralInferShapes<ONNXClipV11Op>(
```

Comment:
This does not work because we expect each operation to have their `computeShape` to perform the precise computations.

Reply:
Because onnx.Cilp op is treated as an elementwise unary operation (like LeakyRelu), the logic is integrated into src/Dialect/ONNX/ONNXOps/Math/Elementwise.cpp, and src/Dialect/ONNX/ONNXOps/Math/Clip.cpp is removed.